pkg/driver/node/mounter/pod_mounter_test.go

```diff
@@ -0,0 +1,379 @@
+				// we should duplicate our file descriptor to ensure underlying file description won't
+				// closed once the file descriptor passed to `PodMounter.Mount` closed.
+				fd, err = syscall.Dup(int(devNull.Fd()))
+				assert.NoError(t, err)
+
+				return fd, nil
+			}
+
+			args := mountpoint.ParseArgs([]string{mountpoint.ArgReadOnly})
+
+			mountRes := make(chan error)
+			go func() {
+				err := testCtx.podMounter.Mount(testCtx.ctx, testCtx.bucketName, testCtx.targetPath, credentialprovider.ProvideContext{
+					AuthenticationSource: credentialprovider.AuthenticationSourceDriver,
+					VolumeID:             testCtx.volumeID,
+					PodID:                testCtx.podUID,
+				}, args)
+				if err != nil {
+					log.Println("Mount failed", err)
+				}
+				mountRes <- err
+			}()
+
+			mpPod := createMountpointPod(testCtx)
+			mpPod.run()
+
+			got := mpPod.receiveMountOptions(testCtx.ctx)
+
+			err := <-mountRes
+			assert.NoError(t, err)
+
+			gotFile := os.NewFile(uintptr(got.Fd), "fd")
+			mountertest.AssertSameFile(t, devNull, gotFile)
+			got.Fd = 0
+
+			assert.Equals(t, mountoptions.Options{
+				BucketName: testCtx.bucketName,
+				Args: []string{
+					"--read-only",
+					"--user-agent-prefix=" + mounter.UserAgent(credentialprovider.AuthenticationSourceDriver, testK8sVersion),
+				},
+				Env: envprovider.Default().List(),
+			}, got)
+		})
+
+		t.Run("Waits for Mountpoint Pod", func(t *testing.T) {
+			testCtx := setup(t)
+
+			go func() {
+				// Add delays to each Mountpoint Pod step
+				time.Sleep(100 * time.Millisecond)
+				mpPod := createMountpointPod(testCtx)
+				time.Sleep(100 * time.Millisecond)
+				mpPod.run()
+				time.Sleep(100 * time.Millisecond)
+				mpPod.receiveMountOptions(testCtx.ctx)
+			}()
+
+			err := testCtx.podMounter.Mount(testCtx.ctx, testCtx.bucketName, testCtx.targetPath, credentialprovider.ProvideContext{
+				VolumeID: testCtx.volumeID,
+				PodID:    testCtx.podUID,
+			}, mountpoint.ParseArgs(nil))
+			assert.NoError(t, err)
+		})
+
+		t.Run("Does not duplicate mounts if target is already mounted", func(t *testing.T) {
+			testCtx := setup(t)
+
+			var mountCount atomic.Int32
+
+			testCtx.mountSyscall = func(target string, args mountpoint.Args) (fd int, err error) {
+				mountCount.Add(1)
+				testCtx.mount.Mount("mountpoint-s3", target, "fuse", nil)
+				return int(mountertest.OpenDevNull(t).Fd()), nil
+			}
+
+			go func() {
+				mpPod := createMountpointPod(testCtx)
+				mpPod.run()
+				mpPod.receiveMountOptions(testCtx.ctx)
+			}()
+
+			for i := 0; i < 5; i++ {
+				err := testCtx.podMounter.Mount(testCtx.ctx, testCtx.bucketName, testCtx.targetPath, credentialprovider.ProvideContext{
+					VolumeID: testCtx.volumeID,
+					PodID:    testCtx.podUID,
+				}, mountpoint.ParseArgs(nil))
+				assert.NoError(t, err)
+			}
+
+			assert.Equals(t, int32(1), mountCount.Load())
+		})
+
+		t.Run("Unmounts target if Mountpoint Pod does not receive mount options", func(t *testing.T) {
```

Comment:
This is unclear - we're verifying that we're unmounting if MP pod fails to mount due to not receiving mount options.
What happens if MP is already mounted here?

Reply:
If MP is already mounted first `IsMountPoint` check would return early. So we know this is a new mount, and if we don't hear back from Mountpoint in case of a failure, we assume it's unhealthy, and we unmount the filesystem. The filesystem is not really exposed to the user as we haven't returned from `NodePublishVolume` call.